### DISCUSSION TITLE
How to modfy default soundfont and fluidsynth settings?

### DISCUSSION THREAD

**Discussion by Ekolide** [cardonabits/haxo-hw#49] [Q&A] [answered]
Hi! I got the Haxophone a couple of days and I am very excited to start digging into it. Though I'm not new to Linux I am new to embedded devices. I'm looking to replace the default soundfont with my own on the rPi Zero image, and have manually crawled through much of the filesystem with no luck.

I'm also looking to disable the Fluidsynth reverb, is there somewhere I can modify the Fluidsynth startup arguments?

Thanks!

**Comment by jcard0na** [comment, marked as answer]
Hi @Ekolide,

If you are familiar with Linux, your learning curve will be smooth.  To make changes to the filesystem, you must first disable write protection.  This is enabled by default on the firmware images we release to prevent corrupting the SD card.  You do this using `raspi-config`, see [last step here](https://github.com/cardonabits/haxo-hw/blob/main/docs/development.md#initial-configuration)

Then you can make changes.  The soundfont is a command-line argument to the `haxo001` executable.  This is set in the file [haxo.service](https://github.com/cardonabits/haxo-rs/blob/main/scripts/systemd/haxo.service#L17) that you will find under `/etc/systemd/system`.

To install a different soundfont, just copy it somewhere in the filesystem and update that path.

As for fluidsynth reverb, I don't really know how that is configured.  The (minimal) fluidsynth configuration in `haxo001` happens [here](https://github.com/cardonabits/haxo-rs/blob/main/src/synth.rs).  At the moment, the only command line arguments are the program number and the soundfont file path.

After making your changes, remember to re-enable write protection.

**Reply by Ekolide** [reply]
Hi!

Thanks so much. My SD card does not seem to be write-protected, I've been switching it between my computer and the rPi to make changes to the filesystem. Not good, I know, but my current solution.

I've successfully changed both the notemap and the soundfont. After switching soundfont, the reverb disappeared, so it might be part of the sample and not part of Fluidsynth. Maybe I am misremembering the Fluidsynth documentation stating that reverb is enabled by default? Though I would still like to interact with the Fluidsynth API, perhaps suitably through haxo.service? For example adjusting polyphony, legato mode, and LADSPA plugins at startup.

The reason I'm interested in portamento is partly because I'm hearing a slight click every time I change note. I think the synth envelopes are retriggering very fast when the note changes, and the polyphony of 1 might be adding to the issue. I tried playing with a soundfont with very long attack and decay but I couldn't tell if it made a difference.

Thanks again!

**Reply by jcard0na** [reply]
Excellent.  Thanks for sharing what you discover.  You seem to know more about midi synthesizing than I do.  If you are able to mitigate those clicks, it would be super.

**Reply by Ekolide** [reply]
Fantastic! It is strange because when listening and using to the soundfonts on my computer all is fine. There's a number of things I'd like to experiment with. On the Fluidsynth side that's polyphony, portamento, legato modes, as well as the number of audio buffers. But I would also like to mess with the audio driver settings for buffer size.

For Fluidsynth configuration, I see bright possibilities for the Haxophone to read and pass along a user config file. Similar to passing -f /foo/bar according to the Fluidsynth [UserManual](https://github.com/FluidSynth/fluidsynth/wiki/UserManual). Of course, I don't know if that's a nightmare task to implement.

**Reply by jcard0na** [reply]
Passing a config file should be simple, but it would need to be parsed by the haxo executable and converted into function calls into [rust-fluidsynth](https://github.com/scholtzan/rust-fluidsynth)

Here are some starting pointers:

1.  How to add additional command-line options: https://github.com/cardonabits/haxo-rs/blob/main/src/main.rs#L30-L39C2
2. How to access those options, and pass them to the (libfluid)synth initialization code: https://github.com/cardonabits/haxo-rs/blob/main/src/main.rs#L77
3. How to populate the `settings` struct that will be used to create the synth: https://github.com/cardonabits/haxo-rs/blob/main/src/synth.rs#L14-L35

If you are going to be re-compiling the application, I recommend you do that on a Raspberry Pi 3 or 4, as the Zero is frustratingly slow for development:  a full compilation takes close to an hour.  This compiles all the dependencies.  After that, just recompiling the executable itself, is much faster, but it is still in the order of minutes.  See [this section](https://github.com/cardonabits/haxo-hw/blob/main/docs/development.md#compilation-method-1-native-compilation-on-the-raspi) on how to compile the code.

Let me know you have any questions.
Best,

**Reply by Ekolide** [reply]
I am willing to tinker, make settings, and samples. Though I have written a "Hello World!" program in C once, it's not for me. I'm a musician with other experiences and expertise.

I have looked through the rust-fluidsynth readme, and it links to Fluidsynth v1 documentation on Sourceforge. What is the current Fluidsynth release on the Haxophone? The readme is also clear that it is an unmaintained project, and doesn't fully integrate the Fluidsynth API. How does the future for that look on the Haxophone?

Would you be willing to extend the init configurations for these things? Sending arbitrary CC messages at startup should be the minimum to change portamento, polyphony, and such. I keep playing the Haxophone and while I'm having a lot of fun I would love to see continuous notes working.

**Reply by jcard0na** [reply]
Got it.  Musicians are always appreciated, with or without hacking skills 😄

> Would you be willing to extend the init configurations for these things? Sending arbitrary CC messages at startup should be the minimum to change portamento, polyphony, and such. I keep playing the Haxophone and while I'm having a lot of fun I would love to see continuous notes working.

Sure.  If you can define exactly what initialization parameters we'd need to send, and what would be an intuitive set of command line arguments to set them, I can take a look.  I have never tried portamento, and I decided to disable polyphony given that there is no polyphony on a regular sax.

**Reply by Ekolide** [reply]
Hi again. That sounds great! I suggest we follow the format outlined in the FluidSynth [UserManual](https://github.com/FluidSynth/fluidsynth/wiki/UserManual#midi-messages) section on MIDI messages: `cc channel ctrl value`

Examples of sane MIDI CCs in this format that make sense to initialize:

````
cc 0 005 000 (Portamento Time, resolution of 128ms, set to 0)
cc 0 037 010 (extension of PortamentoTime, set to 10 = 10ms. See PolyMono documentation section 3.8.5.)
cc 0 065 064 (Portamento Switch, set >= 64 might be enabled. See PolyMono documentation section 3.4.8.)
cc 0 068 064 (Legato Switch, set >=64 might be enabled. Same as above)
cc 0 091 000 (Reverb Send Level, set to 0. I don't think this is the same as disabling the reverb, it might be enabled in the background even if not sending anything) 
cc 0 093 000 (Chorus Send Level, set to 0. Same as above)
````

(or maybe they're fine being formatted with single digits, I don't know, but this looks relatively visually clean)

The currently supported MIDI CC values Fluidsynth supports are documented [here](https://github.com/FluidSynth/fluidsynth/wiki/FluidFeatures#midi-control-change-implementation-chart). The problem is that there seem to be more CCs supported under the hood. For example, in the PolyMono docs linked on that page, section 3.8.5. "Portamento time: CC MSB(5d) and LSB(37d)" it mentions modifying CC 37d (which I think might just means CC 37) as a way of getting portamento times lower than 128ms.

For haxo.service, I suggest `--sf2-init-cc`. Then the user can enter a string of `"cc channel ctrl value cc channel ctrl value cc channel ctrl value"` as they need or want. I don't want to take the focus away from this, but in a future I imagine it would be cool for the notemap to send midi cc information like mod wheel, different octaves triggering different programs in the soundfont, or setting sustain on/off.

Nonetheless, I imagine the haxo.service looking a little bit like this. I moved the lines around a bit and I renamed `prog-number` into `sf2-prog-number`, I don't know if that's helpful or just more visual noise. 

````
ExecStart=/usr/local/bin/haxo001 \
          --notemap-file /foo/bar/notemap.json \
          --sf2-file /foo/bar/soundfont.sf2 \
          --sf2-prog-number 0 \
          --sf2-init-cc "cc 0 005 5 cc 0 037 010 cc 0 065 64 cc 0 068 64 cc 0 091 000 cc 0 093 000"
````

It doesn't look _nice_, but it's not reinventing anything, and I imagine it's easy for you to parse.

As for the polyphony, mono makes sense from an acoustic perspective. Sometimes with mono settings in the digital realm, we get these clipping artifacts as the sound engine cuts the previous note on any arbitrary sample, not letting it ring out or crossfade. Hopefully, with making progress on legato, it won't matter.

**Reply by jcard0na** [reply]
Thanks for putting the effort into proposing an interface for this feature and linking to the supporting docs.  It is indeed easy to parse, and the version of `rust-fluidsynth` we use, even if no longer actively developed, [already supports sending Control Change messages](https://github.com/scholtzan/rust-fluidsynth/blob/master/src/synth.rs#L98).  So the implementation should be easy.

Three questions:

1. Given that this is not Sound Font 2 related, is the `sf2` prefix required? i.e would `--init-cc` or even `--synth-init-cc` make more sense?
2. How long do you think that init string can get?  If it is much longer than what you have in your example, we should probably make it a file (e.g. `--init-cc-file`) with each message in a separate line:
```
cc 0 005 5 
cc 0 037 010 
cc 0 065 64 
cc 0 068 64 
cc 0 091 000 
cc 0 093 000
...
```
3. As you have seen, the haxophone only uses channel 0.  Would it be OK to omit the `0`? i.e.
```
cc 005 5 
cc 037 010 
...
```

Cheers,

**Reply by Ekolide** [reply]
At the end of the day the specific implementation of course has to be left up to you. With that said, here are some thoughts.

1. `--synth-init-cc` or `--init-cc-file` could make sense. In practical terms, I imagine that there might be three situations where one wants to set these types of settings: a) global user-config with sane defaults for all sound fonts. b) per Soundfont, so that if the Haxophone ever gets the ability to switch between different Soundfonts, different settings can be applied for each. c) Per program in each Soundfont. This might not be easily doable, but imagine if you have different reverb, chorus, and portamento settings for your bass sound on program 3 than on your melodic flute sound on program 5, for example. This might not be easily implementable or worth doing, on the other hand. My point is, on a technical level, this might be considered part of the Fluidsynth `synth` engine, but from a user perspective, it feels like you're kind of dialing in the settings for the Soundfont. Does that make sense? Maybe it doesn't matter very much for the user that's already this deep in the system?
2. Honestly, I don't know. Hard to imagine anyone would want to manually initialize *every single* CC-able parameter, Fluidsynth already does. On the other hand, I think the system should be ready to deal with it. So if something sane might be close to my suggestion, someone trying to debug something might want to try a substantial number of parameters. I think an init file makes sense, we could place other relevant parameters that deals with the Fluidsynth initializing in the same place. Like gain, polyphony, and enabling/disabling reverb/chorus, maybe a specific notemap goes with a specific Soundfont? What if the Haxophone automatically loads the init-config file of any Soundfont that has an identically named text file in the same directory? So if I have `/foo/bar/soundfont.sf2` and `/foo/bar/soundfont.initconfig` (or whatever) they get loaded together? But If I have `/foo/bar/saxophone.sf2` without an associated text file, the Haxophone defaults to the user global init, stored... somewhere? What was the question again? 😅
3. The Haxophone only uses channel 0? Is this why a soundfont that has a preset with two instruments assigned to it only makes the first of them sound? Or is that something else? More testing is required. Nonetheless, I see an overkill future where you can enter polyphonic modes on the Haxophone, triggering two different sounds simultaneously. From that perspective, keeping the channels make sense. But of course, if the Haxophone is built from the ground up to live on one channel, that's that. Though I do want to encourage you to keep it open, as I imagine different users might need to speak with MIDI gear on specific channels. Say, switch to one channel to write a bass synth sequence, switch to another channel to record a loop on the built-in sounds, switch to another channel to trigger a drum machine sequence, etc.

But these are, just like, my opinion, man. Feel free to shine a light on how to improve on these ideas.
